src/h2_stream_set.erl

```diff
@@ -0,0 +1,847 @@
+-record(
+   idle_stream, {
+     id :: stream_id()
+    }).
+-type idle_stream() :: #idle_stream{}.
+
+%% So a stream can be any of these things. And it will be something
+%% that you can pass back into serveral functions here in this module.
+-type stream() :: active_stream()
+                | closed_stream()
+                | idle_stream().
+-export_type([stream/0]).
+
+%% Set Operations
+-export(
+   [
+    new/1,
+    new_stream/7,
+    get/2,
+    upsert/2,
+    sort/1
+   ]).
+
+%% Accessors
+-export(
+   [
+    queued_data/1,
+    update_data_queue/3,
+    decrement_recv_window/2,
+    recv_window_size/1,
+    response/1,
+    send_window_size/1,
+    increment_send_window_size/2,
+    pid/1,
+    stream_id/1,
+    stream_pid/1,
+    notify_pid/1,
+    type/1,
+    my_active_count/1,
+    their_active_count/1,
+    my_active_streams/1,
+    their_active_streams/1
+   ]
+  ).
+
+-export(
+   [
+    close/3,
+    send_what_we_can/4,
+    update_all_recv_windows/2,
+    update_all_send_windows/2,
+    update_their_max_active/2,
+    update_my_max_active/2
+   ]
+  ).
+
+%% new/1 returns a new stream_set. This is your constructor.
+-spec new(
+        client | server
+       ) -> stream_set().
+new(client) ->
+    #stream_set{
+       type=client,
+       %% I'm a client, so mine are always odd numbered
+       mine=
+           #peer_subset{
+              lowest_stream_id=0,
+              next_available_stream_id=1
+             },
+       %% And theirs are always even
+       theirs=
+           #peer_subset{
+              lowest_stream_id=0,
+              next_available_stream_id=2
+             }
+      };
+new(server) ->
+    #stream_set{
+       type=server,
+       %% I'm a server, so mine are always even
+       mine=
+           #peer_subset{
+              lowest_stream_id=0,
+              next_available_stream_id=2
+             },
+       %% And theirs are always odd.
+       theirs=
+           #peer_subset{
+              lowest_stream_id=0,
+              next_available_stream_id=1
+             }
+      }.
+
+-spec new_stream(
+        StreamId :: stream_id(),
+        NotifyPid :: pid(),
+        CBMod :: module(),
+        Socket :: sock:socket(),
+        InitialSendWindow :: integer(),
+        InitialRecvWindow :: integer(),
+        StreamSet :: stream_set()) ->
+                        stream_set()
+                            | {error, error_code(), closed_stream()}.
+new_stream(
+          StreamId,
+          NotifyPid,
+          CBMod,
+          Socket,
+          InitialSendWindow,
+          InitialRecvWindow,
+          StreamSet) ->
+    PeerSubset = get_peer_subset(StreamId, StreamSet),
+    case PeerSubset#peer_subset.max_active =/= unlimited andalso
+         PeerSubset#peer_subset.active_count >= PeerSubset#peer_subset.max_active
+    of
+        true ->
+            {error, ?REFUSED_STREAM, #closed_stream{id=StreamId}};
+        false ->
+            {ok, Pid} = http2_stream:start_link(
```

Comment:
Maybe? I hate to say it, but the upsert function uses the same logic in this case statement as a guard, so If it's true here, it will also be true for upsert. There's probably a better way to do it.